internal/duckdb/duckdb.go

```diff
@@ -0,0 +1,153 @@
+package duckdb
+
+import (
+	"database/sql"
+	"fmt"
+	"go/types"
+	"log"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/pkg/errors"
+	"golang.org/x/text/transform"
+
+	"bitbucket.org/jatone/genieql"
+	"bitbucket.org/jatone/genieql/astutil"
+	"bitbucket.org/jatone/genieql/columninfo"
+	"bitbucket.org/jatone/genieql/dialects"
+	"bitbucket.org/jatone/genieql/internal/debugx"
+	"bitbucket.org/jatone/genieql/internal/transformx"
+)
+
+// Dialect constant representing the dialect name.
+const Dialect = "duckdb"
+
+// NewDialect creates a duckdb Dialect from the queryer
+func NewDialect(q *sql.DB) genieql.Dialect {
+	return dialectImplementation{db: q}
+}
+
+func init() {
+	maybePanic := func(err error) {
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	maybePanic(dialects.Register(Dialect, dialectFactory{}))
+}
+
+type queryer interface {
+	Query(string, ...interface{}) (*sql.Rows, error)
+}
+
+type dialectFactory struct{}
+
+func (t dialectFactory) Connect(config genieql.Configuration) (_ genieql.Dialect, err error) {
+	db, err := sql.Open("duckdb", config.ConnectionURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to connect to DuckDB: %s", config.ConnectionURL)
+	}
+	return dialectImplementation{db: db}, nil
+}
+
+type dialectImplementation struct {
+	db *sql.DB
+}
+
+func (t dialectImplementation) Insert(n int, offset int, table, conflict string, columns, projection, defaults []string) string {
+	return Insert(n, offset, table, conflict, columns, projection, defaults)
+}
+
+func (t dialectImplementation) Select(table string, columns, predicates []string) string {
+	return Select(table, columns, predicates)
+}
+
+func (t dialectImplementation) Update(table string, columns, predicates, returning []string) string {
+	return Update(table, columns, predicates, returning)
+}
+
+func (t dialectImplementation) Delete(table string, columns, predicates []string) string {
+	return Delete(table, columns, predicates)
+}
+
+func (t dialectImplementation) ColumnValueTransformer() genieql.ColumnTransformer {
+	return &columnValueTransformer{}
+}
+
+func (t dialectImplementation) ColumnNameTransformer(transforms ...transform.Transformer) genieql.ColumnTransformer {
+	return columninfo.NewNameTransformer(
+		transformx.Wrap("\""),
+		transform.Chain(transforms...),
+	)
+}
+
+func (t dialectImplementation) ColumnInformationForTable(d genieql.Driver, table string) ([]genieql.ColumnInfo, error) {
+	const columnInformationQuery = `DESCRIBE %s`
+	return columnInformation(d, t.db, columnInformationQuery, table)
+}
+
+func (t dialectImplementation) ColumnInformationForQuery(d genieql.Driver, query string) ([]genieql.ColumnInfo, error) {
+	const columnInformationQuery = `DESCRIBE genieql_query_columns_table`
```

Comment:
[temporary tables](https://duckdb.org/docs/sql/statements/create_table.html#temporary-tables), hash the query and fmt.Sprintf("gql_%s", uid). simpler won't break if conncurrent queries are run.

this is a minor nit since you're using a transaction everything  here should be safe.